[Merged by Bors] - feat(Topology): obtain new distance structures on pseudo(e)metricspaces

FormulaRabbit81 opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)
The changes here were originally in #29321, and have been reviewed, but split off here for clarity.

FormulaRabbit81 committed:
feat(Topology): obtain new distance structures on pseudo(e)metricspaces

YaelDillies reviewed (approved):
I am happy with this, but I was involved in writing it, so I won't sign it off.

sgouezel commented on Mathlib/Topology/MetricSpace/Basic.lean:
```diff
@@ -190,3 +190,101 @@ instance SeparationQuotient.instMetricSpace {α : Type u} [PseudoMetricSpace α]
+
+/-- Build new pseudoemetric space from an old one where the edistance uniform structure is provably
+(but typically non-definitionally) equal to some given distance structure. Below are convenience
+versions for PseudoMetric, Emetric and Metric spaces. -/
```
The docstring of a theorem will typically appear when hovering above a use of the theorem. So, references to "below" or "above" don't really make sense in this context, and the docstrings should be self-contained. So you could say in this one "We also provide convenience versions for ..." (without the Below), and the other ones should essentially be copies of this one, with the same amount of detail.

FormulaRabbit81 committed:
More appropriate + informative docstrings

Updated comments to clarify the purpose of the functions for building new metric spaces from existing ones, emphasizing the relationship between distance structures.

FormulaRabbit81 committed:
Docstring Line length

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

FormulaRabbit81 committed:
corrected docstring

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

sgouezel commented on Mathlib/Topology/MetricSpace/Basic.lean:
```diff
@@ -190,3 +190,113 @@ instance SeparationQuotient.instMetricSpace {α : Type u} [PseudoMetricSpace α]
+namespace MetricSpace
+variable {X : Type*} (m : MetricSpace X) (d : X → X → ℝ) (hd : d = dist)
+
+/-- Build new metric space from an old one where the distance are provably (but typically
```
```suggestion
/-- Build new metric space from an old one where the distance is provably (but typically
```

sgouezel reviewed (commented):
bors d+

FormulaRabbit81 committed:
typo

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

FormulaRabbit81 committed:
typo

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

FormulaRabbit81 commented:
bors r+
